src/webWorker/webWorkerManager.js

```diff
@@ -0,0 +1,319 @@
+// eslint-disable-next-line
+import dicomMicroscopyImageLoaderWebWorker from './index.worker.js'
+
+// the taskId to assign to the next task added via addTask()
+let nextTaskId = 0
+
+// array of queued tasks sorted with highest priority task first
+const tasks = []
+
+// array of web workers to dispatch decode tasks to
+const webWorkers = []
+
+const defaultConfig = {
+  maxWebWorkers: navigator.hardwareConcurrency || 1,
+  webWorkerTaskPaths: [],
+}
+
+let config
+
+const statistics = {
+  maxWebWorkers: 0,
+  numWebWorkers: 0,
+  numTasksQueued: 0,
+  numTasksExecuting: 0,
+  numTasksCompleted: 0,
+  totalTaskTimeInMS: 0,
+  totalTimeDelayedInMS: 0,
+}
+
+/**
+ * Function to start a task on a web worker
+ */
+function startTaskOnWebWorker() {
+  // return immediately if no decode tasks to do
+  if (!tasks.length) {
+    return
+  }
+
+  // look for a web worker that is ready
+  for (let i = 0; i < webWorkers.length; i++) {
+    if (webWorkers[i].status === 'ready') {
+      // mark it as busy so tasks are not assigned to it
+      webWorkers[i].status = 'busy'
+
+      // get the highest priority task
+      const task = tasks.shift()
+
+      task.start = new Date().getTime()
+
+      // update stats with how long this task was delayed (waiting in queue)
+      const end = new Date().getTime()
+
+      statistics.totalTimeDelayedInMS += end - task.added
+
+      // assign this task to this web worker and send the web worker
+      // a message to execute it
+      webWorkers[i].task = task
+      webWorkers[i].worker.postMessage(
+        {
+          taskType: task.taskType,
+          workerIndex: i,
+          data: task.data
+        },
+        task.transferList
+      )
+      statistics.numTasksExecuting++
+
+      return
+    }
+  }
+
+  // if no available web workers and we haven't started max web workers, start a new one
+  if (webWorkers.length < config.maxWebWorkers) {
+    spawnWebWorker()
+  }
```

Comment:
You are right, they were not terminated, I have added the call in In https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/465b884827b84044f6e05551d9cd1c904e160561 :  now `VolumeImageViewer.cleanup()` runs the `terminateAllWebWorkers` function.

Regarding the `spawnWebWorker`. I advise to leave as it is for both perfomances and memory reasons. Currently the WebWorkers are created only if needed dynamically. 

Allocating all of them at viewer `constructor` or `render`, it would imply to give a N parameter for the number of workers and allocating all them at once. This was my first approach, but during testing I experienced:
1) that this can generate out of memory issue (i.e. with N ~>15). 
2) each worker would need to load the dynamic webpack bundle index.worker.js (each worker is a separate process), which is few hundreds of KB. (NOTE: each worker would also need to load the wasm, but the wasm is loaded actaully the first time `decodeAsync` is called and not at a startup).

Reply:
Response to https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76#issuecomment-1126027888:

Thanks for the explanation! That has improved my understanding a lot.

It's probably cleanest to instantiate and terminate web workers for every viewer instance. This may incur on overhead, but decoupling management of the web worker from viewer instances would probably be much more messy and result in more error-prone applications.